### PR TITLE
[124] Add optional flags to override hostname and username

### PR DESCRIPTION
**Pull request by pavannd1**
## Change Overview

- Add optional flags to override hostname and username
- Add the flags to `snapshot`, `policy` and `server start` commands
- Add a test case in the E2E tests to create and list snapshots with overridden hostname and username

## Pull Request Type

- [x] 🌻 Feature

## Issues

- kopia/kopia#124

## Test Plan

- [x] Manual
- [x] E2E

### CODE REVIEW

**Comment by pavannd1** [comment]
I have a couple questions:
1. Some of the `kopia policy *` commands use `policyTargets` which uses `getHostName` and `getUserName` to find target policy. How to override it for those commands? Should I add the flags to the commands using `policyTargets` and pass them in?

2. Similarly, the `kopia server start` command uses `getHostName` and `getUserName`, does that require overriding as well?

File: cli/command_snapshot_create.go [hunk 1]
```diff
@@ -181,6 +181,10 @@ func getLocalBackupPaths(ctx context.Context, rep *repo.Repository) ([]string, e
 }
 
 func getUserName() string {
+	if len(overrideUserName) > 0 {
+		return overrideUserName
+	}
+
```
Comment (julio-lopez) [inline cli/command_snapshot_create.go]: You may simply want to:
* simply return the `username` variable
* have the previous functions return the default user / host names
* initialize the variables with the default user and host names

So, for example:
```suggestion
	return userName
}

func getDefaultUserName() string {
```

File: cli/snapshot_utils.go [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package cli
+
+import (
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+var (
+	overrideHostName string
```
Comment (julio-lopez) [inline cli/snapshot_utils.go]:
```suggestion
	hostName = getDefaultHostName()
```

File: cli/snapshot_utils.go [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+	overrideUserName string
```
Comment (julio-lopez) [inline cli/snapshot_utils.go]:
```suggestion
	userName = getDefaultUserName()
```

File: cli/command_snapshot_create.go [hunk 4]
```diff
@@ -199,6 +203,10 @@ func getUserName() string {
 }
 
 func getHostName() string {
+	if len(overrideHostName) > 0 {
+		return overrideHostName
+	}
+
```
Comment (julio-lopez) [inline cli/command_snapshot_create.go]: Similarly:
```suggestion
	return hostName
}

func getDefaultHostName() string {
```

**Comment by julio-lopez** [comment]
> I have a couple questions:
> 
> 1. Some of the `kopia policy *` commands use `policyTargets` which uses `getHostName` and `getUserName` to find target policy. How to override it for those commands? Should I add the flags to the commands using `policyTargets` and pass them in?

You already addressed these ^^^

> 2. Similarly, the `kopia server start` command uses `getHostName` and `getUserName`, does that require overriding as well?

Dunno, let me check quickly or we can ask Jarek.

**Comment by julio-lopez** [comment]
> > 2. Similarly, the `kopia server start` command uses `getHostName` and `getUserName`, does that require overriding as well?
> 
> Dunno, let me check quickly or we can ask Jarek.

It makes sense to allow specifying the user and host names there as well.
It seems that the same applies to `policy` commands.

**Pull request merged**